### ISSUE TITLE
Trying to catch the log event when there aren't disruptive actions

### ISSUE THREAD

**Issue by cano-devel** [SpiderLabs/ModSecurity#2168]
**Describe the bug**

I'm integrating libmodsecurity (3.0.3) in an HTTP proxy. I'm using the Cpp API.
I don't detect any log event when a SecRule matches and the action is not disruptive.
Reading the library documentation, I understand that the 'intervention' object is which returns the log event:
"
intervention
Check if ModSecurity has anything to ask to the server. Intervention can generate a log event and/or perform a disruptive action.
"

A field called 'log' exists in this struct. It returns a log string when a disruptive action matches:
disruptive	1	int
log	"[client ] ModSecurity: Access denied with code 403 (phase 1). Matched "Operator `Rx' with parameter "... (unknown length)	char*
pause	0	int
status	403	int
url	0x0	char*

but not when it is not disruptive:
intervention	@0x7ffff37d7fe0	modsecurity::ModSecurityIntervention
disruptive	0	int
log	0x0	char*
pause	0	int
status	200	int
url	0x0	char*

also, I tested with the "setServerLogCb" function thinking that it would log automatically, but I couldn't find any invoke of the function (when is this invoked?). I used it as I show in the following code:
 
       # adding it to the object
	modsec->setServerLogCb(logModsec_cb);
  
	# cb definition
	void Waf::logModsec_cb(void *data, const void *message) {

		Debug::logmsg(LOG_WARNING, "----------");
		
		Debug::logmsg(LOG_WARNING, "Reach the modsecurity log cb");
		Debug::logmsg(LOG_WARNING, "Message data");
		Debug::logmsg(LOG_WARNING, "%s", (const char *)data);

		Debug::logmsg(LOG_WARNING, "Message param");
		Debug::logmsg(LOG_WARNING, "%s", (const char *)message);

		Debug::logmsg(LOG_WARNING, "----------");
	}

**To Reproduce**

I test with the following configuration file:
SecRuleEngine on
SecDebugLog /opt/waf/debug.log
SecDebugLogLevel 9
SecDefaultAction "pass,phase:1"
SecRule REQUEST_URI "logme" "\
        id:1,\
        phase:1,\
        log"

the curl command:
curl http://192.168.100.241:8080/logme

**Expected behavior**

Checking intervention after processing the rules I get the intervention object:

intervention	@0x7ffff37d7fe0	modsecurity::ModSecurityIntervention
	disruptive	0	int
	log	0x0	char*
	pause	0	int
	status	200	int
	url	0x0	char*

I was expecting a disruptive=0 and a log string regarding the match. Or a call to ServerLogCb that prints in the syslog (invoking my own cb function)

**Server:**
 - ModSecurity version 3.0.3
 - OS (and distro): Debian buster

**Logs and dumps**

Output of:
 1. DebugLogs (level 9)

[156932853534.143554] [] [4] Initializing transaction
[156932853534.143554] [] [9] Appending request body: 0 bytes. Limit set to: 0.000000
[156932853534.143554] [] [4] Starting phase URI. (SecRules 0 + 1/2)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Starting phase REQUEST_HEADERS.  (SecRules 1)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] This phase consists of 1 rule(s).
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] (Rule: 1) Executing operator "Rx" with param "logme" against REQUEST_URI.
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] Target value: "/logme HTTP/1.1\x0d\x0aHost: 192.168.100.241:8080\x0d\x0aUser-Agent: curl/7.64.0 (35 characters omitted)" (Variable: REQUEST_URI)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] Matched vars updated.
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

**] [4] Rule returned 1.**
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

**] [9] Running action: log**
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

**] [9] Saving transaction to logs**
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Starting phase REQUEST_BODY. (SecRules 2)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] This phase consists of 0 rule(s).
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] Appending response body: 169 bytes. Limit set to: 0.000000
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Starting phase RESPONSE_HEADERS. (SecRules 3)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] This phase consists of 0 rule(s).
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Starting phase RESPONSE_BODY. (SecRules 4)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Response body is disabled, returning... 2
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [4] Starting phase LOGGING. (SecRules 5)
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [9] This phase consists of 0 rule(s).
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [8] Checking if this request is suitable to be saved as an audit log.
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [8] Checking if this request is relevant to be part of the audit logs.
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [5] Audit log engine was not set.
[156932853534.143554] [/logme HTTP/1.1
Host: 192.168.100.241:8080
User-Agent: curl/7.64.0
Accept: */*

] [8] Request was relevant to be saved. Parts: 4430

Thanks for your time and your work

**Comment by zimmerle** [comment]
Hi @alvarocano-zevenet,

That is indeed the correct behavior. The logging is only generated when there is an intervention. A rule match does not necessarily mean that a log should be generated. It could only increment a variable, for instance... What are you trying to achieve?

**Comment by cano-devel** [comment]
Hi @zimmerle, 

The idea is to get rules that only logging when a rule matches. In fact, configuring de SecRuleEngine=DetectionOnly I haven't got the log intervention.

I miss an option to log and no disrupt the connection.

Thanks for your response

**Comment by zimmerle** [comment]
If you want to get logging for every match, you are going to do a few hacks on the library, as it wasn' t meant to be used that way. The question is: why you need to do that?

**Comment by cano-devel** [comment]
I think it is a way of disabling the WAF and take metrics about the impact that it could have. Else I don't understand the use of the 'log/nolog' actions and the 'SecRuleEngine=DetectionOnly' directive.

**Comment by zimmerle** [comment]
In that case you should rely on the DetectionOnly mode and use regular logging callback functions.

**Comment by cano-devel** [comment]
Hi @zimmerlezen 

I did some tests regarding the log issue and I got this [table](https://docs.google.com/spreadsheets/d/1AjX1MSBUNEcuGsGDSICos6EwtxAleDfVKsSAqq1mKL4/edit?usp=sharing)

The "print log" column shows when libmodsecurtiy is using the event intervention or the callback function to print in the log.

Looking at the table I get the following notes:

- The "log" action SecRule parameter looks that is not doing anything regarding the log decision. 
- When SecRuleEngine=DetectOnly and action=deny I haven't got any log event/callback

As another note, I get a disruptive action when the SecRuleEngine is set as "DetectionOnly". The ruleset that I am using in this case is:

```
SecRuleEngine DetectionOnly
SecDefaultAction "deny,phase:1"
SecRule REQUEST_METHOD "PUT" "id:1,block,log"
```

I used the following curl command to force the rule match:
curl -k -X PUT -H 'Content-Type: application/json' -d '{"lang":"SP"}' http://192.168.100.241:8080/

This is the intervention struct that I get:
```
disruptive	1	int
log	"[client 192.168.0.186] ModSecurity: Access denied with code 403 (phase 1). Matched "Operator `Rx' wi"... (unknown length)	char*
pause	0	int
status	403	int
url	0x0	char*
```

Regards
